standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java

```diff
@@ -77,15 +77,14 @@ public static ValidTxnList createValidTxnListForCleaner(GetOpenTxnsResponse txns
         break;
       }
       if (abortedBits.get(i)) {
-        exceptions[i] = txnId;
+        exceptions[i++] = txnId;
       } else {
         if (isAbortCleanup) {
-          exceptions[i] = txnId;
+          exceptions[i++] = txnId;
         } else {
-          assert false : JavaUtils.txnIdToString(txnId) + " is open and <= hwm:" + highWatermark;
+          assert (minOpenTxn == Long.MAX_VALUE) : JavaUtils.txnIdToString(txnId) + " is open and <= hwm:" + highWatermark;
```

Comment:
I think the fundamental idea behind keeping it as false, is that compaction can't treat writes from an open txn if they are present within the highWatermark. Won't this condition add an open txn to the list which is later used by compaction?

Reply:
- Open txns are not added to the exceptions list when isAbortCleanup=false, only aborts.
- When there are no open write txns and we rely on MinHistoryWriteId, minOpenTxn=Long.MAX_VALUE. 
- `assert` is only executed during the test phase, unless -ea option is provided

Reply:
AFAIK, if an open txn is present and it is not added to the exceptions list or not failed. Then it is treated as a valid txn when fetching the associated write IDs.

Reply:
there shouldn't be any open txns below the highWatermark, Cleaner only checks the committed or aborted writes
````
      return new ValidReadTxnList(exceptions, bitSet, highWatermark, Long.MAX_VALUE);
    } else {
      return new ValidReadTxnList(exceptions, abortedBits, highWatermark, Long.MAX_VALUE);
````
it's set to Long.MAX_VALUE

Reply:
i think we need to cap hwm by NEXT_WRITE_ID-1, WDYT?

Reply:
Yes, now it makes sense.